lib/inventory/inventory.go

```diff
@@ -87,12 +87,19 @@ type DownstreamSender interface {
 func NewDownstreamHandle(fn DownstreamCreateFunc, hello proto.UpstreamInventoryHello) DownstreamHandle {
 	ctx, cancel := context.WithCancel(context.Background())
 	handle := &downstreamHandle{
-		senderC:      make(chan DownstreamSender),
-		pingHandlers: make(map[uint64]DownstreamPingHandler),
-		closeContext: ctx,
-		cancel:       cancel,
+		senderC:           make(chan DownstreamSender),
+		pingHandlers:      make(map[uint64]DownstreamPingHandler),
+		closeContext:      ctx,
+		cancel:            cancel,
+		agentMetadataDone: make(chan struct{}),
 	}
 	go handle.run(fn, hello)
+	go func() {
+		handle.agentMetadata = fetchAgentMetadata(&fetchConfig{ctx: ctx, hello: hello})
+		// Signal that the agentMetadata has been calculated by closing the
+		// agentMetadataDone channel.
+		close(handle.agentMetadataDone)
+	}()
```

Comment:
This is more of a stylistic comment, so take it or leave it...

I think this code might be more maintainable and easier to reason about if we tried to keep it as loosely-coupled as possible.  Rather than integrating the metadata fetching/sending into the core `DownstreamHandle` logic, I'd like to propose an alternative that should achieve the same effect (example assumes we also get rid of the need to re-send hello fields):


```golang
func AutoEmitMetadata(handle DownstreamHandle) {
    metadata := fetchAgentMetadata(handle.CloseContext())
    for {
        select {
        case sender := <-handle.Sender(): // wait for stream to be opened
            sender.Send(handle.CloseContext(), metadata) // send metadata
            select {
            case <-sender.Done(): // block for the duration of the stream
            case <-handle.CloseContext.Done():
                return
            }
        case <-handle.CloseContext.Done():
            return
        }
    }
}
```

The above will have the effect of ensuring that metadata is sent once each time a stream is established, but it has the nice effect of being very loosely coupled with the internals of the handle.

This is already how control-stream based heartbeats work (except they send periodically for the life of the stream), and I think it's a good pattern to stick to if possible.  I think `lib/inventory` is complex enough, that any time we can break out components to be separate that's a win.

Reply:
Done in cbdefedc8040014de974167c680ca23dd49c9052.